cdap-common/src/main/java/io/cdap/cdap/common/io/CachingLocation.java

```diff
@@ -0,0 +1,233 @@
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URI;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.util.List;
+import java.util.Objects;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link Location} implementation that caches data read locally to allow efficient re-reading.
+ */
+final class CachingLocation implements Location {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CachingLocation.class);
+
+  private final LocationFactory locationFactory;
+  private final Location delegate;
+  private final CachingPathProvider cachingPathProvider;
+
+  CachingLocation(LocationFactory locationFactory, Location delegate, CachingPathProvider cachingPathProvider) {
+    this.locationFactory = locationFactory;
+    this.delegate = delegate;
+    this.cachingPathProvider = cachingPathProvider;
+  }
+
+  @Override
+  public boolean exists() throws IOException {
+    return delegate.exists();
+  }
+
+  @Override
+  public String getName() {
+    return delegate.getName();
+  }
+
+  @Override
+  public boolean createNew() throws IOException {
+    return delegate.createNew();
+  }
+
+  @Override
+  public boolean createNew(String permission) throws IOException {
+    return delegate.createNew(permission);
+  }
+
+  @Override
+  public String getPermissions() throws IOException {
+    return delegate.getPermissions();
+  }
+
+  @Override
+  public String getOwner() throws IOException {
+    return delegate.getOwner();
+  }
+
+  @Override
+  public String getGroup() throws IOException {
+    return delegate.getGroup();
+  }
+
+  @Override
+  public void setGroup(String group) throws IOException {
+    delegate.setGroup(group);
+  }
+
+  @Override
+  public void setPermissions(String permission) throws IOException {
+    delegate.setPermissions(permission);
+  }
+
+  @Override
+  public InputStream getInputStream() throws IOException {
+    Path cachePath = cachingPathProvider.apply(delegate).orElse(null);
+    if (cachePath == null) {
+      return delegate.getInputStream();
+    }
+
+    try {
+      return new FileInputStream(cachePath.toFile());
+    } catch (IOException e) {
+      // If fail to open stream from the cache, try to regenerate it.
+      Files.deleteIfExists(cachePath);
+    }
+
+    try (InputStream input = delegate.getInputStream()) {
+      Files.createDirectories(cachePath.getParent());
+      Path tmpPath = Files.createTempFile(cachePath.getParent(), cachePath.getFileName().toString(), ".tmp",
+                                          PosixFilePermissions.asFileAttribute(
+                                            PosixFilePermissions.fromString("rw-------")));
+      Files.copy(input, tmpPath, StandardCopyOption.REPLACE_EXISTING);
```

Comment:
do we need to delete the file at the tmpPath? Or is that somehow handled by Files.createTempFile?

Reply:
It is handled by `StandardCopyOption.REPLACE_EXISTING`.